### DISCUSSION TITLE
Pwsh is interferring with PowerShell ability to find Get-FileHash

### DISCUSSION THREAD

**Discussion by jaredpar** [PowerShell/PowerShell#24630] [General]
Recently run into a problem with using `Get-FileHash` as it seems to be unavailable in some contexts. After a lot of experimentation it seems that when `pwsh` indirectly launches `powershell` it interferes with its ability to find functions like `Get-FileHash`. To demonstrate this consider the following files:

test.ps1
```ps1
$h = Get-FileHash "test.ps1" 
Write-Host $h.Hash
```

test-powershell.cmd
```cmd
@echo off
powershell -noprofile -file "%~dp0\test.ps1" %*
```

test-pwsh.cmd
```cmd
@echo off
pwsh -noprofile -file "%~dp0\test.ps1" %*
```

Now run the following under both a cmd and pwsh shell. 

```sh
> test-pwsh.cmd
> test-powershell.cmd
```

What I observe is that when running both from a cmd shell everything works as expected. The `Get-FileHash` function is available and prints out the file hash. But when I run from a pwsh shell the `Get-FileHash` function is not available. 

```txt
C:\Users\jaredpar\temp\ps
⚡🔨 > .\test-pwsh.cmd
C44C543F8829387C3AC9E1955EB96D645AC5EBFF35E58DF4624AC9428B68891B
C:\Users\jaredpar\temp\ps
⚡🔨 > .\test-powershell.cmd
Get-FileHash : The term 'Get-FileHash' is not recognized as the name of a cmdlet, function, script file,
or operable program. Check the spelling of the name, or if a path was included, verify that the path is
correct and try again.
At C:\Users\jaredpar\temp\ps\test.ps1:1 char:6
+ $h = Get-FileHash "test.ps1"
+      ~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (Get-FileHash:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException
```

I cannot make any sense of why launching from a pwsh shell has this impact on the code here. What am I missing? 

This scenario is important because a non-trivial amount of our infrastructure needs to run on Powershell. Yet a lot of our dev team runs pwsh shells locally. This behavior means that we cannot run our infra locally because we end up hitting errors like this. Yes I can fix locally by just flipping to a cmd shell but it took me an hour to even root cause pwsh being the issue here.

| PS Kind | Version |
| --- | --- |
| powershell | 5.1.26100.2161 |
| pwsh | 7.4.6 |

Operating system: windows

**Comment by iSazonov** [comment]
pwsh has implemented side-by-side support with PowerShell. This is not the truth in the opposite direction. In other words, push knows that PowerShell exists, but PowerShell does not know that pwsh exists. The problem is that they _both_ use the _same_ PSModulePath env to search for modules.
There's a multi-year dance with PSModulePath on the pwsh side that's still going on. 
If you restore PSModulePath to its original state before running PowerShell, then your script will work.

**Reply by iSazonov** [reply]
I have two comments.
1. You can run bat in cmd.exe where to explicitly set this variable
2. It is generally assumed that the user always runs scripts in pwsh. If there is a module that is not supported in pwsh, then it will automatically be launched in PowerShell in the background (Compatibility feature) by remoting (serialization!). If this does not work, then you should open the case in this repository.

**Reply by jaredpar** [reply]
Thanks for the help! 

Think the way I'm going to approach this is to just reset `%PSMODULEPATH%` in the .cmd files

```cmd
@echo off
set PSMODULEPATH=
powershell -noprofile -file "%~dp0\test.ps1" %*
```

That will make it resilient to this type of problem.

[a comment by an automated account is not shown]
